addons/mail/static/src/widgets/form_renderer/form_renderer_tests.js

```diff
@@ -853,6 +853,89 @@ QUnit.test('Form view not scrolled when switching record', async function (asser
     );
 });
 
+QUnit.test('Attachments that have been unlinked from server should be visually unlinked from record', async function (assert) {
+    // Attachments that have been fetched from a record at certain time and then
+    // removed from the server should be reflected on the UI when the current
+    // partner accesses this record again.
+    assert.expect(2);
+
+    this.data['res.partner'].records.push(
+        { display_name: "Partner1", id: 11 },
+        { display_name: "Partner2", id: 12 }
+    );
+
+    this.data['ir.attachment'].records.push(
+        {
+           id: 11,
+           mimetype: 'text.txt',
+           res_id: 11,
+           res_model: 'res.partner',
+        },
+        {
+           id: 12,
+           mimetype: 'text.txt',
+           res_id: 11,
+           res_model: 'res.partner',
+        }
+    );
+
+    await this.createView({
+        data: this.data,
+        hasView: true,
+        // View params
+        View: FormView,
+        model: 'res.partner',
+        res_id: 11,
+        viewOptions: {
+            ids: [11, 12],
+            index: 0,
+        },
+        arch: `
+            <form string="Partners">
+                <sheet>
+                    <field name="name"/>
+                </sheet>
+                <div class="oe_chatter">
+                    <field name="message_ids"/>
+                </div>
+            </form>
+        `,
+    });
+
+    await afterNextRender(() =>
+        document.querySelector('.o_ChatterTopbar_buttonAttachments').click()
+    );
+
+    assert.containsN(
+        document.body,
+        '.o_AttachmentList_attachment',
+        2,
+        "There should be two attachments displayed"
+    );
+
+    await afterNextRender(() =>
+        document.querySelector('.o_pager_next').click()
+    );
+
+    // Simulate unlinking attachment 12 from Partner 1.
+    this.data['ir.attachment'].records.find(a => a.id === 11).res_id = 12;
```

Comment:
Why not just remove the attachment from data? I think it's safer to test removal of attachment than change of link.

A reason to prefer that is because a change of link in server are not immediately visible on UI.
The test simulates "unlink" of an attachment from record "Partner1" that is not visible, by being in "Partner2".
Problem with changing link to "Partner2" is that in context of test, would expect displaying 1 attachment in "Partner2", but it's not.

Sure, it only happens if someone debugs test and puts breakpoint right after this simulate, but this could be confusing the person debugging the test. Plain removal of attachment is easier to understand and there's no weird other question to ask ourself.